### DISCUSSION TITLE
How to implement a tag named paginate like  shopiffy?

### DISCUSSION THREAD

**Discussion by wuyonggang100** [harttle/liquidjs#403] [Working with Shopify]
https://shopify.dev/api/liquid/objects/paginate
https://shopify.dev/api/liquid/tags/theme-tags#layout

**Comment by harttle** [comment]
We do have a [layout](https://liquidjs.com/tags/layout.html) tag to make LiquidJS functionable out of the box. But other Shopify tags/filters will only be added into liquidjs if they're not Shopify business specific. Otherwise creating a plugin to maintain a given set of filgers/tags is recommended.

Feel free to open an issue for liquid support while doing so, or ask a question in [LiquidJS Discussion](https://github.com/harttle/liquidjs/discussions).

**Comment by wuyonggang100** [comment]
the question is how to get these variable like this in a plugin ? 
`import { assert, Tokenizer, evalToken, Emitter, TagToken, TopLevelToken, Context, Template, TagImplOptions, ParseStream } from '../../types'`

**Comment by ericraio** [comment]
In ruby liquid, I was able to create the paginate block by inheriting from Liquid::Block. For liquidjs, I am trying to figure out how to accomplish the same thing and it looks like the necessary components are hidden away and not exported.

**Comment by harttle** [comment]
@wuyonggang100 names in `types.ts` are exported by [liquid.ts](https://github.com/harttle/liquidjs/blob/4f6b88c154154cfda869a69188ae5682039d49c0/src/liquid.ts#L18), can be obtained by importing liquidjs. Like:

```typescript
import {TagToken} from 'liquidjs'
```

**Comment by harttle** [comment]
@ericraio not sure what `Block` does. According to [this doc](https://rubydoc.info/gems/liquid/Liquid/Block), I suppose it's something like Tag or classes used by Tag. We do have [registerTags](https://liquidjs.com/tutorials/register-filters-tags.html) and `Drop`, the latter one is not well documented but I believe it's usable.

* registerTags: https://liquidjs.com/tutorials/register-filters-tags.html
* Drop: not well documented, but
  * A sample usage can be found here: https://github.com/harttle/liquidjs/pull/398/files
  * Types doc can be found here: https://liquidjs.com/api/classes/drop_drop_.drop.html

**Comment by ericraio** [comment]
What I am looking to accomplish is creating my own for loop tag where in
shopify you can use the syntax like

{% paginate collection.products by 20 %}
{% endpaginate %}

In ruby, I accomplished this syntax from inheriting from Liquid::Block

On Mon, Oct 4, 2021 at 6:45 AM Jun Yang ***@***.***> wrote:

> @ericraio <https://github.com/ericraio> not sure what Block does.
> According to this doc <https://rubydoc.info/gems/liquid/Liquid/Block>, I
> suppose it's something like Tag or classes used by Tag. We do have
> registerTags <https://liquidjs.com/tutorials/register-filters-tags.html>
> and Drop, the latter one is not well documented but I believe it's usable.
>
>    - registerTags:
>    https://liquidjs.com/tutorials/register-filters-tags.html
>    - Drop: not well documented, but
>       - A sample usage can be found here:
>       https://github.com/harttle/liquidjs/pull/398/files
>       - Types doc can be found here:
>       https://liquidjs.com/api/classes/drop_drop_.drop.html
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/harttle/liquidjs/issues/387#issuecomment-933502062>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAKW6G36IN4ARMRJEOZFWTUFGVXVANCNFSM5D43SL3Q>
> .
>


-- 
- Eric Raio

**Comment by harttle** [comment]
Seems like just Tag, implement something like [if](https://github.com/harttle/liquidjs/blob/master/src/builtin/tags/if.ts) and register using `registerTag` will do the job.

**Comment by ericraio** [comment]
@harttle can tags perform iterations ? that's what shopify's block is doing

**Reply by harttle** [reply]
You can refer to https://github.com/harttle/liquidjs/tree/master/src/builtin/tags for examples to implement tags.

**Reply by ericraio** [reply]
Thank you @harttle , I got around to implementing this and took a moment to figure this out

For others who want to create this.

You need to copy into your project ForloopDrop, toEnumerable function and the for loop tag. At the top of the copied over for loop tag, my imports look like this.

```
  const { toEnumerable } = require("./util/collection");
  const ForloopDrop = require("./forloop-drop");
  const { assert, Tokenizer, evalToken, Hash } = require("liquidjs");
  ```
  
  Then with the copied over for loop, you'll have to step through it to convert it to match shopify's paginate.
